### PR TITLE
Process all science exposures for a given night and tile

### PR DESCRIPTION
**Pull request by marcelo-alvarez**
Adds a new script `bin/desi_proc_tilenight` that uses functions defined  in `desispec.scripts.proc_tilenight`, to process all science exposures for a given night and tile combination -- i.e. the prestdstar, stdstar, and poststdstar steps in the spectroscopic pipeline. 

Testing on Perlmutter was carried out on a single interactive node, requested with 
```
salloc --qos=interactive -N 1 --time=120 -C gpu -A desi_g --gpus-per-node=4
```
for runs labeled with `GPU` (running with `--gpuspecter` and `--gpuextract`), and 
```
salloc --qos=interactive -N 1 --time=120 -C cpu -A desi
```
for runs labeled with `CPU` (without `--gpuextract`), as follows:
```
source /global/common/software/desi/desi_environment.sh main
DESI_SPECTRO_REDUX=/global/cfs/cdirs/desi/users/$USER/spectro/redux

branchdir=$SCRATCH/desispec-tilenight
git clone https://github.com/desihub/desispec --branch tilenight $branchdir
PYTHONPATH=$branchdir/py:$PYTHONPATH
PATH=$branchdir/bin:$PATH

export OMP_NUM_THREADS=1
export MPICH_GPU_SUPPORT_ENABLED=1 # only set for perlmutter-gpu 

night=20210203
tileid=80699

preprod_ngt () {
  proddir=$DESI_SPECTRO_REDUX/$SPECPROD
  mkdir -p $proddir
  cd $proddir
  if [ ! -d $proddir/exposure_tables/${night:0:6} ] ; then 
    copyprod $CFS/desi/spectro/redux/$cpspec . --night $night --filter-exptables
  fi
  rm -rf $proddir/exposures/$night $proddir/preproc/$night
}
logfile=tilenight_$night\_$tileid.log

# gpu_specter on GPUs
SPECPROD=tilenight-gpuext-gpuspc
preprod_ngt
srun -n $nproc -c 2 --cpu-bind=cores desi_mps_wrapper desi_proc_tilenight -n $night -t $tileid --mpistdstars --mpi --gpuextract --gpuspecter |& tee -a $logfile

# gpu_specter on CPUs 
SPECPROD=tilenight-cpuext-gpuspc
preprod_ngt
srun -n $nproc -c 2 --cpu-bind=cores desi_proc_tilenight -n $night -t $tileid --mpistdstars --mpi  --gpuspecter |& tee $logfile

# specter on CPUs 
SPECPROD=tilenight-cpuext-cpuspc
preprod_ngt
srun -n $nproc -c 2 --cpu-bind=cores desi_proc_tilenight -n $night -t $tileid --mpistdstars --mpi   |& tee $logfile
```
with the following timings for each:
```
specter      extraction   extraction    total wall  nproc       partition
version            type         time    clock time
                         (53 frames)

gpu_specter        GPU    442s  (7m)    803s (13m)     32  perlmutter-gpu
specter            CPU   3122s (52m)   3160s (58m)     32  perlmutter-gpu
specter            CPU    714s (12m)   1063s (17m)    128  perlmutter-cpu

gpu_specter        GPU    448s  (7m)    800s (13m)     64  perlmutter-gpu
gpu_specter        CPU    657s (11m)    986s (16m)    128  perlmutter-cpu
```
Input welcome from @dmargala and @akremin, particularly on gpu_specter performance and pipeline integration, respectively.

### CODE REVIEW

**Comment by sbailey** [comment]
Thanks!  At a first glance, this looks good, though I will try it myself and review in more detail.  As a quick note, for Fuji on KNL this tilenight took 5 jobs and 305 node-minutes, which is now down to one job and 13.3 node-minutes on Perlmutter GPU (!).

I suspect the CPU numbers could be better, since the Perlmutter CPU nodes have 128 physical cores, and even the GPU nodes have 64 physical cores, but your test runs above only used 32 MPI ranks.  Some steps can't use all the cores, but please also run "specter on CPUs" timing tests on Perlmutter CPU nodes with "srun -n 128 ...", and also "gpu_specter on GPUs" using "srun -n 64 ...".  Thanks.

**Comment by marcelo-alvarez** [comment]
> 
> I suspect the CPU numbers could be better, since the Perlmutter CPU nodes have 128 physical cores, and even the GPU nodes have 64 physical cores, but your test runs above only used 32 MPI ranks. Some steps can't use all the cores, but please also run "specter on CPUs" timing tests on Perlmutter CPU nodes with "srun -n 128 ...", and also "gpu_specter on GPUs" using "srun -n 64 ...". Thanks.
@sbailey I did three additional runs (left out the 32 processor gpu_specter CPU job as it would be very similar to to 32 processor specter CPU). 

These results indicate a speedup of using a GPU accelerated node over a CPU node, for the 53-frame tilenight used here, of about 3-4 minutes, cutting down from 16 min to 13 min, or about a 20-25% performance increase. Depending on the state of the system and the charge for GPU node hours vs CPU node hours, it may very well be better to run tilenight on CPU-only nodes in the current configuration.

**Review by akremin — COMMENTED** [review]
This looks great! I added some comments and a few very minor change requests inline. I'll post more details after this on the `badamps` comment and the `stdstar` comment.

**Comment by akremin** [comment]
A BADAMPS example:
On night 20211014,  two amplifiers had issues: r1D and r4A for all tiles. One example tile is 6699 with exposures 104277, 104278, and 104279.

The commands we want to see are in these Slurm files: 
```
desi@cori21 [daily]$ ls $DESI_SPECTRO_REDUX/himalayas/run/scripts/night/20211014/*104277*.slurm
/global/cfs/cdirs/desi/spectro/redux/himalayas/run/scripts/night/20211014/poststdstar-20211014-00104277-a0123456789.slurm 
/global/cfs/cdirs/desi/spectro/redux/himalayas/run/scripts/night/20211014/stdstarfit-20211014-00104277-a0123456789.slurm
/global/cfs/cdirs/desi/spectro/redux/himalayas/run/scripts/night/20211014/prestdstar-20211014-00104277-a0123456789.slurm
```

Below I give the commands passed to srun. As I think you know the script generator adds the` --timingfile`, `--starttime`, and `--mpi` tags. I don't know if you have an equivalent script generator or if you need to add them at the function level.

An example prestdstar is:
```
desi_proc  --nostdstarfit --nofluxcalib --cameras a0123456789 -n 20211014 -e 104277 --badamps r1D,r4A --mpi --starttime $(date +%s) --timingfile /global/cfs/cdirs/desi/spectro/redux/himalayas/run/scripts/night/20211014/prestdstar-20211014-00104277-a0123456789-timing-$SLURM_JOBID.json
```

Example stdstar (unchanged from what you should have):
```
desi_proc_joint_fit  --obstype science --cameras a0123456789 -n 20211014 -e 104277,104278,104279 --mpi --mpistdstars --starttime $(date +%s) --timingfile /global/cfs/cdirs/desi/spectro/redux/himalayas/run/scripts/night/20211014/stdstarfit-20211014-00104277-a0123456789-timing-$SLURM_JOBID.json
```

Example poststdstar. I don't think the badamp here are strictly necessary, but they don't hurt anything:
``` 
desi_proc  --noprestdstarfit --nostdstarfit --cameras a0123456789 -n 20211014 -e 104277 --badamps r1D,r4A --mpi --starttime $(date +%s) --timingfile /global/cfs/cdirs/desi/spectro/redux/himalayas/run/scripts/night/20211014/poststdstar-20211014-00104277-a0123456789-timing-$SLURM_JOBID.json
```

**Comment by akremin** [comment]
Finally a discussion on `stdstar` fitting, which we should bring @julienguy in for:

This script would break from that convention and run all exposures of a tile on a night, even if one was at the beginning of the night and two were taken late in the night. 

In the past we only fit `stdsstar`'s on consecutively acquired exposures. I don't recall the exact arguments, but it may have had to do with concern about modeling the sky over longer periods of the night. Another possibility is the absorption feature varying with humidity and conditions throughout the night. We now mitigate both of these concerns to a larger degree than we used to, but I don't know if they are sufficiently solved to no longer worry about the joint fitting.

I spoke with Stephen and he is now in favor of the choice made for this new script.

Question to Julien -- do you think there is any problem with combining all exposures of a tile for a night, even if they are spread out across the night?

File: py/desispec/scripts/proc_tilenight.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    #- What are we going to do?
+    if rank == 0:
+        log.info('----------')
+        log.info('Tile {} night {}'.format(args.tileid, args.night))
+        log.info('Output root {}'.format(specprod_root()))
+        log.info('----------')
+
+    #- Determine expids and cameras for a tile night
+    reduxdir = specprod_root()
+    exptable_file = f'{reduxdir}/exposure_tables/{str(args.night)[0:6]}/exposure_table_{args.night}.csv'
```
[the remark this answers is not in the document]
Comment (marcelo-alvarez) [reply py/desispec/scripts/proc_tilenight.py]: used, this, thanks.

File: py/desispec/scripts/proc_tilenight.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    keep  = exptable['OBSTYPE'] == 'science'
+    keep &= exptable['TILEID']  == int(args.tileid)
+    exptable = exptable[keep]
+
+    expids = list(exptable['EXPID'])
+    prestdstar_expids = []
+    stdstar_expids = []
+    poststdstar_expids = []
+    cameras = dict()
```
[the remark this answers is not in the document]
Comment (marcelo-alvarez) [reply py/desispec/scripts/proc_tilenight.py]: done.

**Comment by marcelo-alvarez** [comment]
> A BADAMPS example: On night 20211014, two amplifiers had issues: r1D and r4A for all tiles. One example tile is 6699 with exposures 104277, 104278, and 104279.

@kremin, thanks for this, I've added BADAMPS processing and verified the correct arguments are passed for this example while leaving processing of exposures for which the BADAMPS entry is blank unchanged.

> 
> As I think you know the script generator adds the` --timingfile`, `--starttime`, and `--mpi` tags. I don't know if you have an equivalent script generator or if you need to add them at the function level.

We will address automated script generation separately from this PR.

File: py/desispec/scripts/proc_tilenight.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+    #- Determine expids and cameras for a tile night
+    reduxdir = specprod_root()
+    exptable_file = f'{reduxdir}/exposure_tables/{str(args.night)[0:6]}/exposure_table_{args.night}.csv'
+    log.info(f'Reading exptable in {exptable_file}')
+
+    exptable = load_table(exptable_file)
+
+    keep  = exptable['OBSTYPE'] == 'science'
+    keep &= exptable['TILEID']  == int(args.tileid)
```
[the remark this answers is not in the document]
Comment (marcelo-alvarez) [reply py/desispec/scripts/proc_tilenight.py:65]: Ok, marking this resolved pending discussion about this below and elsewhere, thanks.

File: py/desispec/scripts/proc_tilenight.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    prestdstar_expids = []
+    stdstar_expids = []
+    poststdstar_expids = []
+    cameras = dict()
+    for i in range(len(expids)):
+        expid=expids[i]
+        camword=exptable['CAMWORD'][i]
+        badcamword=exptable['BADCAMWORD'][i]
+        if isinstance(badcamword, str):
+            full_cameras = decode_camword(camword)
```
[the remark this answers is not in the document]
Comment (akremin) [reply py/desispec/scripts/proc_tilenight.py:80]: Perhaps we add an argument to optionally suppress? We would keep the default to do logging to maintain backward compatibility, but you could suppress them here.

In other words: 
`def difference_camwords(fullcamword,badcamword)` -> def `difference_camwords(fullcamword,badcamword, suppress_logging=False)`

and you use `suppress_logging=True` in this script.

**Comment by marcelo-alvarez** [comment]
> Finally a discussion on `stdstar` fitting, which we should bring @julienguy in for:
> 
> This script would break from that convention and run all exposures of a tile on a night, even if one was at the beginning of the night and two were taken late in the night.
> 
> In the past we only fit `stdsstar`'s on consecutively acquired exposures. I don't recall the exact arguments, but it may have had to do with concern about modeling the sky over longer periods of the night. Another possibility is the absorption feature varying with humidity and conditions throughout the night. We now mitigate both of these concerns to a larger degree than we used to, but I don't know if they are sufficiently solved to no longer worry about the joint fitting.
> 
> I spoke with Stephen and he is now in favor of the choice made for this new script.
> 
> Question to Julien -- do you think there is any problem with combining all exposures of a tile for a night, even if they are spread out across the night?

@kremin one possibility if we wish to retain the ability to do either mode of fitting is to add a flag `--contiguous`. Then one would call `proc_joint_fit` once for each contiguously acquired set of exposures. I would need your input on how to extract this information from the exposure table, but it seems like a reasonable feature to have.

**Comment by sbailey** [comment]
@akremin @julienguy and I just chatted about the non-contiguous observations and agreed that it is ok, even desirable, to fit all exposures of the tile on the night even if they aren't back-to-back, i.e. as this code is currently implemented.  No need to add a --contiguous flag to support the current pipeline behavior for these rare cases.

I think that covers everything needed for this PR.  I'll take one last look and then merge later today.

**Pull request merged**